### ISSUE TITLE
[Core] Prevent schedulling non-GPU tasks to GPU nodes

### ISSUE THREAD

**Issue by eduardohenriquearnold** [ray-project/ray#47866]
## Context
I have a GPU node pool, which defaults to 0 active nodes in order to save compute resources. 
When I submit tasks that require a GPU, that node pool is scaled on demand via k8s/RayCluster.
Once that task finishes, another task which doesn’t require a GPU starts, but it’s being scheduled to the currently active GPU node. Ideally, I would like this follow-up task to use or to spin-up a non-GPU node, allowing the GPU node to be de-allocated. 

## Feature Request
Be able to restrict tasks that do not require a GPU (`num_gpus=0`) to non-GPU nodes, i.e. nodes in which `num_gpus=0`.

## What I have tried
I have considered a workaround which is creating a custom resource, e.g. `no_gpu`, and assigning it to `1` for all nodes that do not have a GPU, then requesting `no_gpu=0.1` for such non-GPU tasks. That will force Ray to schedule such tasks to these nodes.

The problem with this approach is that the nodes with custom resources need to be active for the RayHead node to know about the custom resources' existence. So, for example, if I have some high-memory pools which have 0 default active nodes, Ray doesn't "know" that these nodes need to be spun-up in order to create `no_gpu` resources. AFAIK, the custom resource definition happens in the command-line of starting the Ray worker, so the head has no way of knowing that it needs to scale a given pool in order to increase a specific custom resource.

### Use case

Given a cluster with the following pools:
1. GPU pool (defaults to 0 active nodes)
2. High memory pool  (defaults to 0 active nodes)
3. Regular CPU pool (1 active node by default, RayHead)

I want to be able to submit two sequential tasks:

```Task A (requires 1 GPU) -> Task B (requires 0 GPU, requires high mem), -> End```

and get the first task to spin-up and run on the GPU node pool, and the second task to run on the high-mem pool, freeing resources from the GPU pool to either run other tasks that do require the GPU or to get de-allocated.

Note that:
*  I cannot set node affinity for tasks to run on a specific node, since the nodes do not yet exist, they'll be spun up automatically based on demand.
* I cannot set a specific threshold of memory of Task B such that it surpass the GPU node available memory and Ray is forced to schedule it in the high-mem pool. The reason for this constraint is that I may want to run multiple instances of Task B in a single high-mem node.

**Comment by anyscalesam** [comment]
Great descriptive in the task definition @eduardohenriquearnold > we'll do our weekly triage next Monday and have our sprint planning probably the week after (that should take our sprint to ~3rd week of November).

Stay tuned~

**Comment by jjyao** [comment]
@eduardohenriquearnold Are you using kuberay or cluster launcher? You are able to tell autoscaler that certain node pool has `no_gpu` custom resource.

**Comment by eduardohenriquearnold** [comment]
@jjyao thanks for your attention to this issue. I am using kuberay. Is there a specific documentation/example on how can I tell the autoscaler about custom node pool resources?

**Comment by anyscalesam** [comment]
I think it's just the ray.ScalingConfig.num_gpus var that you set - keep me honest here @jjyao

**Comment by eduardohenriquearnold** [comment]
@anyscalesam I believe that sets the `num_gpus` resource, not a custom resource of any kind? Here we need a custom resource, which in my example I called `no_gpu`, but could be anything really.

**Comment by antoniomdk** [comment]
It'd be great if Ray could support the K8s concept of taints and tolerations. In K8s, with the nvidia-plugin all GPU nodes are tainted so only workloads that request a GPU can be allocated there. The same applies to other accelerators via the [ExtendedResourceToleration](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#extendedresourcetoleration) admission controller.

I was wondering, if in Kuberay when the worker group scales from 0, does the autoscaler know all the available custom resources ahead of time, or just the custom resources available to the cluster at a given time?

**Comment by anyscalesam** [comment]
@antoniomdk the Ray Autoscaler or the K8s one?

**Comment by sofam** [comment]
> It'd be great if Ray could support the K8s concept of taints and tolerations. In K8s, with the nvidia-plugin all GPU nodes are tainted so only workloads that request a GPU can be allocated there. The same applies to other accelerators via the [ExtendedResourceToleration](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#extendedresourcetoleration) admission controller.
> 
> I was wondering, if in Kuberay when the worker group scales from 0, does the autoscaler know all the available custom resources ahead of time, or just the custom resources available to the cluster at a given time?

No, that's one of the issues. The custom resources are only known when the worker group is scaled up. And yes, seconded on this entire thing. It would be great if we could do this because right now running a heterogenous cluster with CPU and GPU nodes mixed, Ray Data internals (for example for metadata fetching) will scale up _all_ of the CPUs in the cluster. You can set the placement group on Ray Data to do the Parquet fetching on a certain shape of node, but the metadata fetcher and other internals will do its own scheduling.

**Comment by jjyao** [comment]
@eduardohenriquearnold see the doc here: https://docs.ray.io/en/latest/cluster/kubernetes/user-guides/config.html#id1

**Comment by eduardohenriquearnold** [comment]
@jjyao I believe this specifies how to create the custom resources within the `rayStartParams` key for RayCluster. As I mentioned in the original thread, this is not enough, as the RayHead/Autoscaler does not know that this specific node pool needs to be scaled up to get such resources. This is also precisely what @sofam described [above](https://github.com/ray-project/ray/issues/47866#issuecomment-2411293569).

**Comment by jjyao** [comment]
@eduardohenriquearnold per the doc

This field can be used to specify custom resource capacities for the Ray pod. These resource capacities will be advertised to the Ray scheduler and **Ray autoscaler**. For example, the following annotation will mark a Ray pod as having 1 unit of Custom1 capacity and 5 units of Custom2 capacity.

Autoscaler should know which node pool to scale up. Otherwise, it's a bug.

**Comment by juliuslipp** [comment]
@jjyao Seems to be a bug.

The json deserialisation is different for ray start and the autoscaler. There's no option where you can pass a valid string that both work at the same time. F.e. this only works for the autoscaler but not for ray start: https://docs.ray.io/en/latest/cluster/kubernetes/user-guides/config.html#id1

@eduardohenriquearnold did you find any workaround?

https://github.com/ray-project/ray/blob/7a372c74d9385d7163c2cfed2de001632fe19810/python/ray/_private/utils.py#L1435-L1454

and

https://github.com/ray-project/ray/blob/717cafb5245b5ee70cec87b92aaab873b0683c4c/python/ray/autoscaler/_private/kuberay/autoscaling_config.py#L403-436

**Comment by eduardohenriquearnold** [comment]
Thanks for digging into this @juliuslipp . Unfortunately, I haven't found a workaround as this issue got de-prioritized from my todos.

**Comment by Oblynx** [comment]
I was curious if there's been any progress here? This unexpected scheduling behavior is causing us costly incidents. Since it was de-prioritized, maybe more accounts of its impact in production can help it get some attention back.

**Comment by Tom-Newton** [comment]
I'm also interested in this problem. I've seen cases where it blocks scaling:
Non-GPU actors get scheduled on a GPU worker -> there are not enough CPU resources to fit a GPU actor on that worker -> no GPU work is done by the GPU worker -> low GPU utilisation -> ray data cluster autoscaler doesn't trigger scale up -> my workload takes ages running with 1 GPU actor.
